libs/wire-api/src/Wire/API/MakesFederatedCall.hs

```diff
@@ -0,0 +1,121 @@
+import Servant.Client
+import Servant.Server
+import Servant.Swagger
+import Test.QuickCheck (Arbitrary)
+import Unsafe.Coerce (unsafeCoerce)
+import Wire.Arbitrary (GenericUniform (..))
+
+data Component
+  = Brig
+  | Galley
+  | Cargohold
+  deriving (Show, Eq, Generic)
+  deriving (Arbitrary) via (GenericUniform Component)
+
+-- | A typeclass corresponding to calls to federated services. This class has
+-- no methods, and exists only to automatically propagate information up to
+-- servant.
+--
+-- The only way to discharge this constraint is via 'callsFed', which should be
+-- invoked for each federated call when connecting handlers to the server
+-- definition.
+class CallsFed (comp :: Component) (name :: Symbol)
+
+-- | A typeclass with the same layout as 'CallsFed', which exists only so we
+-- can discharge 'CallsFeds' constraints by unsafely coercing this one.
+class Nullary
+instance Nullary
+
+-- | Construct a dictionary for 'CallsFed'.
+synthesizeCallsFed :: forall (comp :: Component) (name :: Symbol). Dict (CallsFed comp name)
+synthesizeCallsFed = unsafeCoerce $ Dict @Nullary
+
+-- | Servant combinator for tracking calls to federated calls. Annotating API
+-- endpoints with 'MakesFederatedCall' is the only way to eliminate 'CallsFed'
+-- constraints on handlers.
+data MakesFederatedCall (comp :: Component) (name :: Symbol)
+
+instance (HasServer api ctx) => HasServer (MakesFederatedCall comp name :> api :: *) ctx where
+  -- | This should have type @CallsFed comp name => ServerT api m@, but GHC
+  -- complains loudly thinking this is a polytype. We need to introduce the
+  -- 'CallsFed' constraint so that we can eliminate it via
+  -- 'synthesizeCallsFed', which otherwise is too-high rank for GHC to notice
+  -- we've solved our constraint.
+  type ServerT (MakesFederatedCall comp name :> api) m = Dict (CallsFed comp name) -> ServerT api m
+  route _ ctx f = route (Proxy @api) ctx $ fmap ($ synthesizeCallsFed @comp @name) f
+  hoistServerWithContext _ ctx f s = hoistServerWithContext (Proxy @api) ctx f . s
+
+instance HasLink api => HasLink (MakesFederatedCall comp name :> api :: *) where
+  type MkLink (MakesFederatedCall comp name :> api) x = MkLink api x
+  toLink f _ l = toLink f (Proxy @api) l
+
+instance RoutesToPaths api => RoutesToPaths (MakesFederatedCall comp name :> api :: *) where
+  getRoutes = getRoutes @api
+
+-- | Get a symbol representation of our component.
+type family ShowComponent (x :: Component) :: Symbol where
+  ShowComponent 'Brig = "brig"
+  ShowComponent 'Galley = "galley"
+  ShowComponent 'Cargohold = "cargohold"
+
+-- | 'MakesFederatedCall' annotates the swagger documentation with an extension
+-- tag @x-wire-makes-federated-calls-to@.
+instance (HasSwagger api, KnownSymbol name, KnownSymbol (ShowComponent comp)) => HasSwagger (MakesFederatedCall comp name :> api :: *) where
+  toSwagger _ =
+    toSwagger (Proxy @api)
+      & applyTags
+        [ Tag
+            "x-wire-makes-federated-call-to"
```

Comment:
i just built this branch and tried this out, but came up empty:

```
$ curl -q http://localhost:8082/api/swagger.json | jq . | grep x-wire-makes-federated-call-to
```

am i holding it wrong?

Reply:
I'm not sure how to run the server, but dumping the output of `Wire.API.Routes.Internal.Brig.swaggerDoc` gives me the following output:

[swagger.json.txt](https://github.com/isovector/wire-server/files/10280571/swagger.json.txt)

which gives the following output

```
$ cat /tmp/swagger.json.txt| jq . | grep x-wire-makes-federated-call-to
          "x-wire-makes-federated-call-to"
          "x-wire-makes-federated-call-to"
      "name": "x-wire-makes-federated-call-to"
```

But upon viewing this in a swagger viewer, it looks like tags might be the wrong place for this information:

![2022-12-21-112522_585x281_scrot](https://user-images.githubusercontent.com/307223/208987036-ca764e72-e5d1-42e0-bb9b-d83e23b550d1.png)